File: engine/simulation/success-calculator.ts
import { ActionType, Position } from "@/engine/enums";
import type { Player, PlayerAttributes } from "@/engine/models";
import type { IRng } from "./rng";

/**
 * Calculates success chance for a given action.
 * Formula: success_chance = attribute + chemistry_bonus + morale_bonus + trait_bonus + rng_value
 */

/**
 * Calculates the total success chance and compares against a threshold.
 * Returns true if the action succeeds, false otherwise.
 */
export function calculate(
  player: Player,
  action: ActionType,
  assignedPosition: Position,
  rng: IRng,
  bonusModifier: number = 0,
): boolean {
  const successChance =
    calculateRaw(player, action, assignedPosition, rng) + bonusModifier;
  const threshold = getThreshold(action);
  return successChance >= threshold;
}

/**
 * Returns the raw success value (before threshold comparison).
 * Useful for testing and debugging.
 */
export function calculateRaw(
  player: Player,
  action: ActionType,
  assignedPosition: Position,
  rng: IRng,
): number {
  const attribute = getPrimaryAttribute(player.attributes, action);
  const chemistryBonus = getChemistryBonus(player.chemistry);
  const moraleBonus = getMoraleBonus(player.morale);
  const traitBonus = getTraitBonus(player.traits, action);
  const rngValue = rng.nextFloat(-15, 15);
  const positionPenalty = getPositionPenalty(
    player.primaryPosition,
    player.secondaryPosition,
    assignedPosition,
  );

  return (
    attribute * (1 - positionPenalty) +
    chemistryBonus +
    moraleBonus +
    traitBonus +
    rngValue
  );
}

/** Maps an action to its primary attribute. */
export function getPrimaryAttribute(
  attributes: PlayerAttributes,
  action: ActionType,
): number {
  switch (action) {
    case ActionType.Pass:
      return attributes.passing;
    case ActionType.Dribble:
      return attributes.dribbling;
    case ActionType.Shot:
      return attributes.finishing;
    case ActionType.Cross:
      return attributes.technique;
    case ActionType.Tackle:
      return attributes.strength;
    case ActionType.Interception:
      return attributes.anticipation;
    default:
      throw new Error(`Unknown action type: ${action}`);
  }
}

/** Chemistry bonus: 0-20 mapped from chemistry 0-100. */
export function getChemistryBonus(chemistry: number): number {
  return (chemistry / 100) * 20;
}

/** Morale bonus: -10 to +10 mapped from morale 0-100 (50 = neutral). */
export function getMoraleBonus(morale: number): number {
  return ((morale - 50) / 50) * 10;
}

/** Trait bonus: checks if any of the player's traits match the action context. Returns 0-15. */
export function getTraitBonus(
  traits: readonly string[],
  action: ActionType,
): number {
  let bonus = 0;

  for (const trait of traits) {
    bonus += getSingleTraitBonus(trait, action);
  }

  return Math.min(bonus, 15);
}

/**
 * Position penalty based on how far the assigned position is from the player's natural positions.
 * Primary: 0%, Secondary: 10%, Related: 20%, Out of position: 35%.
 */
export function getPositionPenalty(
  primary: Position,
  secondary: Position | null,
  assigned: Position,
): number {
  if (assigned === primary) return 0;
  if (secondary != null && assigned === secondary) return 0.1;
  if (arePositionsRelated(primary, assigned)) return 0.2;
  return 0.35;
}

/** Thresholds vary by action -- shots are hardest, passes are easiest. */
export function getThreshold(action: ActionType): number {
  switch (action) {
    case ActionType.Pass:
      return 40;
    case ActionType.Dribble:
      return 55;
    case ActionType.Shot:
      return 85;
    case ActionType.Cross:
      return 50;
    case ActionType.Tackle:
      return 45;
    case ActionType.Interception:
      return 45;
    default:
      return 50;
  }
}

function getSingleTraitBonus(trait: string, action: ActionType): number {
  if (trait === "Finesse Shot" && action === ActionType.Shot) return 8;
  if (trait === "Power Shot" && action === ActionType.Shot) return 10;
  if (trait === "Close Control" && action === ActionType.Dribble) return 8;
  if (trait === "Through Pass" && action === ActionType.Pass) return 8;
  if (trait === "Interceptor" && action === ActionType.Interception) return 10;
  if (trait === "Aerial Dominance" && action === ActionType.Tackle) return 5;
  if (trait === "Clinical Finisher" && action === ActionType.Shot) return 12;
  if (trait === "Playmaker" && action === ActionType.Pass) return 7;
  if (trait === "Hard Tackler" && action === ActionType.Tackle) return 8;
  if (trait === "Whipped Crosses" && action === ActionType.Cross) return 8;
  return 0;
}

/** Determines if two positions are in the same general area (adapted position = -20%). */
function arePositionsRelated(a: Position, b: Position): boolean {
  return getPositionGroup(a) === getPositionGroup(b);
}

function getPositionGroup(position: Position): number {
  switch (position) {
    case Position.GK:
      return 0;
    case Position.CB:
    case Position.LB:
    case Position.RB:
    case Position.LWB:
    case Position.RWB:
      return 1;
    case Position.CDM:
    case Position.CM:
    case Position.CAM:
    case Position.LM:
    case Position.RM:
      return 2;
    case Position.LW:
    case Position.RW:
    case Position.CF:
    case Position.ST:
      return 3;
    default:
      return -1;
  }
}

File: src/simulation/SuccessCalculator.cs
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;

namespace ElevenLegends.Simulation;

/// <summary>
/// Calculates success chance for a given action.
/// Formula: success_chance = attribute + chemistry_bonus + morale_bonus + trait_bonus + rng_value
/// </summary>
public static class SuccessCalculator
{
    /// <summary>
    /// Calculates the total success chance and compares against a threshold.
    /// </summary>
    /// <param name="player">The player performing the action.</param>
    /// <param name="action">The type of action being attempted.</param>
    /// <param name="assignedPosition">The position the player is currently playing.</param>
    /// <param name="rng">Injected RNG for determinism.</param>
    /// <returns>True if the action succeeds, false otherwise.</returns>
    public static bool Calculate(
        Player player, ActionType action, Position assignedPosition, IRng rng,
        float bonusModifier = 0f)
    {
        float successChance = CalculateRaw(player, action, assignedPosition, rng) + bonusModifier;
        float threshold = GetThreshold(action);
        return successChance >= threshold;
    }

    /// <summary>
    /// Returns the raw success value (before threshold comparison).
    /// Useful for testing and debugging.
    /// </summary>
    public static float CalculateRaw(Player player, ActionType action, Position assignedPosition, IRng rng)
    {
        int attribute = GetPrimaryAttribute(player.Attributes, action);
        float chemistryBonus = GetChemistryBonus(player.Chemistry);
        float moraleBonus = GetMoraleBonus(player.Morale);
        float traitBonus = GetTraitBonus(player.Traits, action);
        float rngValue = rng.NextFloat(-15f, 15f);
        float positionPenalty = GetPositionPenalty(player.PrimaryPosition, player.SecondaryPosition, assignedPosition);

        return (attribute * (1f - positionPenalty)) + chemistryBonus + moraleBonus + traitBonus + rngValue;
    }

    /// <summary>
    /// Maps an action to its primary attribute.
    /// </summary>
    public static int GetPrimaryAttribute(PlayerAttributes attributes, ActionType action)
    {
        return action switch
        {
            ActionType.Pass => attributes.Passing,
            ActionType.Dribble => attributes.Dribbling,
            ActionType.Shot => attributes.Finishing,
            ActionType.Cross => attributes.Technique,
            ActionType.Tackle => attributes.Strength,
            ActionType.Interception => attributes.Anticipation,
            _ => throw new ArgumentOutOfRangeException(nameof(action))
        };
    }

    /// <summary>
    /// Chemistry bonus: 0–20 mapped from chemistry 0–100.
    /// </summary>
    public static float GetChemistryBonus(int chemistry)
    {
        return chemistry / 100f * 20f;
    }

    /// <summary>
    /// Morale bonus: -10 to +10 mapped from morale 0–100 (50 = neutral).
    /// </summary>
    public static float GetMoraleBonus(int morale)
    {
        return (morale - 50f) / 50f * 10f;
    }

    /// <summary>
    /// Trait bonus: checks if any of the player's traits match the action context.
    /// Returns 0–15.
    /// </summary>
    public static float GetTraitBonus(IReadOnlyList<string> traits, ActionType action)
    {
        float bonus = 0f;

        foreach (string trait in traits)
        {
            bonus += GetSingleTraitBonus(trait, action);
        }

        return Math.Min(bonus, 15f);
    }

    /// <summary>
    /// Position penalty based on how far the assigned position is from the player's natural positions.
    /// Primary: 0%, Secondary: 10%, Out of position: 35%.
    /// </summary>
    public static float GetPositionPenalty(Position primary, Position? secondary, Position assigned)
    {
        if (assigned == primary) return 0f;
        if (secondary.HasValue && assigned == secondary.Value) return 0.10f;
        if (ArePositionsRelated(primary, assigned)) return 0.20f;
        return 0.35f;
    }

    /// <summary>
    /// Thresholds vary by action — shots are hardest, passes are easiest.
    /// </summary>
    public static float GetThreshold(ActionType action)
    {
        return action switch
        {
            ActionType.Pass => 40f,
            ActionType.Dribble => 55f,
            ActionType.Shot => 85f,
            ActionType.Cross => 50f,
            ActionType.Tackle => 45f,
            ActionType.Interception => 45f,
            _ => 50f
        };
    }

    private static float GetSingleTraitBonus(string trait, ActionType action)
    {
        return (trait, action) switch
        {
            ("Finesse Shot", ActionType.Shot) => 8f,
            ("Power Shot", ActionType.Shot) => 10f,
            ("Close Control", ActionType.Dribble) => 8f,
            ("Through Pass", ActionType.Pass) => 8f,
            ("Interceptor", ActionType.Interception) => 10f,
            ("Aerial Dominance", ActionType.Tackle) => 5f,
            ("Clinical Finisher", ActionType.Shot) => 12f,
            ("Playmaker", ActionType.Pass) => 7f,
            ("Hard Tackler", ActionType.Tackle) => 8f,
            ("Whipped Crosses", ActionType.Cross) => 8f,
            _ => 0f
        };
    }

    /// <summary>
    /// Determines if two positions are in the same general area (adapted position = -20%).
    /// </summary>
    private static bool ArePositionsRelated(Position a, Position b)
    {
        return GetPositionGroup(a) == GetPositionGroup(b);
    }

    private static int GetPositionGroup(Position position)
    {
        return position switch
        {
            Position.GK => 0,
            Position.CB or Position.LB or Position.RB or Position.LWB or Position.RWB => 1,
            Position.CDM or Position.CM or Position.CAM or Position.LM or Position.RM => 2,
            Position.LW or Position.RW or Position.CF or Position.ST => 3,
            _ => -1
        };
    }
}
